### PR TITLE
DEP: adjust deprecation of positional arguments

### PR DESCRIPTION
**Pull request by j-bowhay**
#### Reference issue

https://github.com/scipy/scipy/issues/18578#issuecomment-1569500078
#### What does this implement/fix?

In the past, when deprecating an argument to a function that we want to remove, we have used the argument being different from the default value of `None` to trigger the deprecation warning. However, as pointed out by @rgommers, the current method we are using to deprecate arguments we want to remove was not entirely watertight if the user uses all positional arguments.

As an example, `kendalltau(x, y, None, 'omit', 'exact')` would not raise the deprecation warning so we are in danger of breaking the user's code without warning if this was removed directly. Side note, this is yet another reason why making new code use keyword-only arguments when applicable is desirable.

As per @rgommers suggestion, I have added a private `_NoValue` object which is used to trigger the deprecation warning. This means now:
```
In [2]: kendalltau([1,4], [1,2], None, 'omit', 'exact')
<ipython-input-2-e55511017c59>:1: DeprecationWarning: 'kendalltau' keyword argument 'initial_lexsort' is deprecated as it is unused and will be removed in SciPy 1.12.0.
  kendalltau([1,4], [1,2], None, 'omit', 'exact')
```
as required to remove the argument.
#### Additional information

cc @h-vetinari

### CODE REVIEW

File: scipy/integrate/_quadrature.py [hunk 1]
```diff
@@ -531,7 +532,7 @@ def simps(y, x=None, dx=1.0, axis=-1, even=None):
     return simpson(y, x=x, dx=dx, axis=axis, even=even)
 
 
-def simpson(y, x=None, dx=1.0, axis=-1, even=None):
+def simpson(y, x=None, dx=1.0, axis=-1, even=_NoValue):
```
Comment (h-vetinari) [inline scipy/integrate/_quadrature.py:535]: This is how things would look like after switching to keyword-only args.
```suggestion
def simpson(y, *, x=None, dx=1.0, axis=-1):
```
In the meantime, we should probably vendor the [helper](https://github.com/scikit-learn/scikit-learn/blob/1.2.2/sklearn/utils/validation.py#L42) that scikit-learn came up with to do the deprecation of positional use.

File: scipy/integrate/_quadrature.py [hunk 2]
```diff
@@ -522,7 +523,7 @@ def _basic_simpson(y, start, stop, x, dx, axis):
 
 # Note: alias kept for backwards compatibility. simps was renamed to simpson
 # because the former is a slur in colloquial English (see gh-12924).
-def simps(y, x=None, dx=1.0, axis=-1, even=None):
+def simps(y, x=None, dx=1.0, axis=-1, even=_NoValue):
     """An alias of `simpson`.
 
     `simps` is kept for backwards compatibility. For new code, prefer
```
Comment (h-vetinari) [inline scipy/integrate/_quadrature.py:529]: Sidenote: this alias sounds like it should eventually be removed. Some digging revealed it came from https://github.com/scipy/scipy/issues/12924, making it even more of a candidate for deprecation and removal.

**Review by h-vetinari — COMMENTED** [review]
If we're extending the deprecation by two releases, I'd like to fully deprecate positional use of all kwargs along the lines of #14714. Example below.

**Comment by rgommers** [comment]
> If we're extending the deprecation by two releases, I'd like to fully deprecate positional use of all kwargs along the lines of #14714. Example below.

It seems to me like this is trying to reopen the gh-14714 discussion, and it's not really related to the topic of this PR - which is only about ensuring warnings for already deprecated keywords are always emitted. I'd suggest not mixing new deprecations into this PR.

File: scipy/signal/_fir_filter_design.py [hunk 3]
```diff
@@ -20,13 +21,13 @@ def _get_fs(fs, nyq):
     """
     Utility for replacing the argument 'nyq' (with default 1) with 'fs'.
     """
-    if nyq is None and fs is None:
+    if nyq is _NoValue and fs is None:
```
Comment (rgommers) [inline scipy/signal/_fir_filter_design.py:24]: It looks to me like this should handle None as well, e.g.:
```python
if nyq in (_NoValue, None) and fs is None:
```

File: scipy/integrate/_quadrature.py [hunk 4]
```diff
@@ -669,7 +670,7 @@ def simpson(y, x=None, dx=1.0, axis=-1, even=None):
         slice_all = (slice(None),) * nd
 
         # default is 'simpson'
-        even = even if even is not None else "simpson"
+        even = even if even is not _NoValue else "simpson"
```
Comment (rgommers) [inline scipy/integrate/_quadrature.py:673]: This should also check if `even` is not `None`, e.g.:
```python
even = even if even is not in (_NoValue, None) else "simpson"
```
or alternatively, undo the change in this line and add this right below the deprecation warning block:
```python
else:
    even = None
```

**Review by rgommers — COMMENTED** [review]
Thanks @j-bowhay. This LGTM module two cases where I think an explicitly given `None` is now no longer handled.

**Comment by j-bowhay** [comment]
> If we're extending the deprecation by two releases, I'd like to fully deprecate positional use of all kwargs along the lines of #14714. Example below.

I'm not sure I follow, this would increase the number of people affected by the deprecation **massively** regardless of their use of the deprecated behaviour.

**Comment by h-vetinari** [comment]
I'm not talking about all functions, I'm talking about all positional arguments in the function being changed here. As soon as we start touching the (positional) signature, we should switch the respective options to keyword-only.

**Comment by rgommers** [comment]
I'd still say that that is orthogonal to this PR, and can/should be done separately. It can be done for the next release for some or all functions in this PR if you really want, but I'm ambivalent about it at best.

**Comment by h-vetinari** [comment]
> I'd still say that that is orthogonal to this PR, and can/should be done separately

This PR already makes positional usage of all arguments following the ones defaulting to `_NoValue` effectively impossible (unless users hack around and provide `_NoValue` positionally, which is an effort I doubt anyone would do). For example:
```python
def kendalltau(x, y, initial_lexsort=_NoValue, nan_policy='propagate',
               method='auto', variant='b', alternative='two-sided'):
```
will always raise for positional usage of any argument past `y`, and therefore already implicitly does what I propose to do explicitly (and I'd say, in a more self-explanatory way).

> but I'm ambivalent about it at best.

The discussion in #14714 basically boils down to (please correct me if you have a different impression): it's an improvement, but not worth it for a wholesale change, however it _is_ beneficial to adopt when we're already changing the signature of a function (as we'll be doing here once we remove the `_NoValue` arguments).

File: scipy/integrate/_quadrature.py (hunk 1 above)
Comment (tupui) [reply scipy/integrate/_quadrature.py:535]: I am +1000 there!

**Comment by rgommers** [comment]
> The discussion in #14714 basically boils down to (please correct me if you have a different impression): it's an improvement, but not worth it for a wholesale change, however it _is_ beneficial to adopt when we're already changing the signature of a function (as we'll be doing here once we remove the `_NoValue` arguments).

Yeah that's about right I think. That said, I think this PR does not make such changes, it only makes pre-existing deprecations more correct. Hence I'm saying it's an orthogonal change and should not block merging this PR.

**Comment by j-bowhay** [comment]
I'm on holiday so I won't be making any changes for at least a week but it seems the way forward is to implement @rgommers suggestions and then look at @h-vetinari's in a follow-up pr.

**Comment by h-vetinari** [comment]
> That said, I think this PR does not make such changes

If you look at the `kendalltau` example I gave, replacing the default with one that's (intentionally) not user-accessible is exactly equivalent to deprecating positional arguments (i.e. there no way[^1] to keep `x, y, None, ...` positional), only that the warning here hides that fact rather than highlight it. So I don't think it's orthogonal.

[^1]: up to importing the private `_NoValue` and using that, but that's hardly API usage we want to encourage

**Comment by h-vetinari** [comment]
> It's true for `kendalltau` only because it's the first keyword that was deprecated

I understand that; positional use becomes impossible exactly for kwargs following those now defaulting to `_NoValue`. However, my main point here is that the message "kwarg X is deprecated" does _not_ communicate, much less explain, that all following positional arguments need to be changed (how? why?). I think this has the potential for substantial frustration, hence why I'd like to be clear about that upfront.

While we're still touching the signature of `simpson` (the bar that was set for #14714), I'm fine if we exclude functions where the `_NoValue` default is for the _last_ argument. In that case we can do the change to kwarg-only separately, but all the functions where we're already breaking positional use (of any non-deprecated argument) should just be deprecated and changed once.

**Comment by rgommers** [comment]
> However, my main point here is that the message "kwarg X is deprecated" does _not_ communicate, much less explain, that all following positional arguments need to be changed

It's the only way to get rid of the deprecation warning, this PR in its current state is perfectly fine. There is no room for confusion here.

**Comment by h-vetinari** [comment]
> There is no room for confusion here.

That's a stretch, given how there's a very plausible scenario for confusion (positional use that now becomes impossible), especially given that the warning does not guide people to the right solution (i.e. "please use keyword arguments").

But before we continue going in circles, I'll drop my concerns for this particular PR, and I can open another issue for 1.12 that targets this topic.

**Comment by ilayn** [comment]
I couldn't follow the discussion here but could someone let me know what the implication would be for #18488 ?

**Comment by rgommers** [comment]
> I couldn't follow the discussion here but could someone let me know what the implication would be for #18488 ?

You're deprecating the `tol` keyword there, right? You want to be using the same `_NoValue` pattern as in this PR. Right now, I believe in your branch, something like `bicg(A, b, x0, None, 100)` won't raise a deprecation warning (because it's gated by `tol is not None`), and then once `tol` gets removed that code breaks. The purpose of the `_NoValue` pattern is to distinguish between "user doesn't use keyword" and "user uses keyword explicitly with its default value"

**Comment by ilayn** [comment]
OK, that means, I'll wait for a deprecation cycle starting from 1.12 to 1.14 if anybody sets tol with positional arg which is your example. Then 1.14 rips `tol` out. But in the meantime am I also putting `*` for strict kwarging? Because I want to :) since there are like 4 different deprecations going on simultaneously in the iterative solvers. I got a bit tired of them just by looking. 

That was the part I didn't parse in this discussion.

**Comment by h-vetinari** [comment]
> But in the meantime am I also putting `*` for strict kwarging?

That's gonna need an additional deprecation for those people feeding the current kwargs positionally. This is what the above-promised issue will be about, because doing that deprecation is more involved (if starting from scratch, which is why we should build on the work from scikit-learn on this).

So no `*` just yet, but also for `1.14` once the depreciation expires.

**Comment by rgommers** [comment]
> So no `*` just yet, but also for `1.14` once the depreciation expires.

This is right.

> That's gonna need an additional deprecation for those people feeding the current kwargs positionally.

It may not, depending on the details - in @ilayn's case I think this does not require an additional deprecation since `tol` is already the first keyword of the ones of interest (I suspect leaving `x0` as positional is fine/desired, because it's not a meaningful name). Current signature:
```python
bicg(A, b, x0=None, tol=None, maxiter=None, M=None, callback=None, atol=0., rtol=1e-5)
```
New signature for 1.14.0 with only the deprecation that's already there plus the `_NoValue` pattern:
```python
# You may also reorder the keywords at that point if you desire to do so for clarity
bicg(A, b, x0=None, *, maxiter=None, M=None, callback=None, atol=0., rtol=1e-5)
```

There's of course more issues (e.g. `M` is a very bad name); if you'd have the freedom to start from scratch it'd be more like:
```python
# Order these keywords from most to least used
bicg(A, b, /, *, guess=None, atol=0., rtol=1e-5, maxiter=None, preconditioner=None, callback=None)

```

> ...  (i.e. "please use keyword arguments").

@h-vetinari perhaps you can follow up with a minor concrete tweak to @ilayn's deprecation message? I suspect it may be annoying to get that right (because you don't know when it applies without more introspection) and it may not be worth it, but please do try to make it concrete, because it'll be obvious once you actually try.

**Comment by ilayn** [comment]
Ah OK it's much clearer for me now too. This PR looks indeed then ready to go, and we can do some follow ups right after this. 1.12 looks like the big clean-up with all the fortran, signatures, and deprecations and so on :) Let me take the discussion to #18488 then and what it would take to mark it as down.

**Comment by rgommers** [comment]
Sounds good to me!

**Comment by ilayn** [comment]
@j-bowhay Oops, I created a conflict, sorry about that. In the meantime, I'm preparing another PR for iterative kwargs having `_NoObject`. Should I wait for you or get the declaration in the new PR?

**Comment by j-bowhay** [comment]
@ilayn I have hopefully addressed all @rgommers comments so this should be good to go if it comes back green. I ended up removing any of the changes to the iterative solver as I thought it would be easier to add it to your pr and do them all at once rather than hunt through the merge conflicts for my changes

File: scipy/integrate/_quadrature.py (hunk 2 above)
Comment (h-vetinari) [reply scipy/integrate/_quadrature.py:529]: https://github.com/scipy/scipy/issues/18702

**Comment by h-vetinari** [comment]
> This is what the above-promised issue will be about

https://github.com/scipy/scipy/issues/18703

**Comment by ilayn** [comment]
Just to confirm, is everyone happy to click the button about this?

**Comment by ilayn** [comment]
All thumbs up from the participants and thanks all for the discussions. It is quite informative for me too.
